### PR TITLE
Random forest

### PR DESCRIPTION
**Pull request by wagenrace**
Work in progress

The following pull request adds, fixes, and/or modifies:

Adding notebook to do a naive random forest without optimization
New requirements:
- seaborn

Still to-do:
- [x] add readme with results
- [x] add futher steps plan
- [x] add analysis pipeline step

I performed the following prior to filing this pull request:

- [x] Tested that my change does not break the analysis pipeline
- [x] Ran a linter through my code
- [x] Update environment dependencies if my code introduces a new package

### CODE REVIEW

File: 3.random_forest/random_forest.ipynb [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+     },
+     "execution_count": 13,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "test_file = os.path.join(data_dir, 'test_processed.tsv.gz')\n",
+    "test_data = pd.read_csv(test_file, sep='\\t')\n",
```
Comment (mth-gsk) [inline 3.random_forest/random_forest.ipynb]: perhaps use `pd.read_table` instead of `read_csv`?

File: 2.process-data/split_training_test.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+test_traget_count = pd.DataFrame(test_df.target.value_counts())
+
+target_count_df = train_traget_count.merge(
+    test_traget_count, left_index=True, right_index=True
+)
+
+target_count_df.columns = ["train_counts", "test_counts"]
+
+file = os.path.join(result_dir, "target_counts.tsv")
+target_count_df.to_csv(file, sep="\t", index=True)
```
Comment (mth-gsk) [inline 2.process-data/split_training_test.py]: Any reason for TSV over CSV?

File: 2.process-data/split_training_test.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+# Load full data set
+full_train_df = pd.read_csv(train_file)
+print(f"Shape of full data set {full_train_df.shape}")
+
+train_df, test_df = train_test_split(
+    full_train_df, test_size=0.1, stratify=full_train_df.target
+)
+
+# Save trainings set
+file = os.path.join(output_data_dir, "train.tsv.gz")
```
Comment (mth-gsk) [inline 2.process-data/split_training_test.py]: avoid using `file` as a variable name; use `filepath` instead.

File: 2.process-data/split_training_test.py (hunk 2 above)
Comment (wagenrace) [reply 2.process-data/split_training_test.py]: They look nice if you visualize them. Look at /2.process-data/results/target_counts.tsv
But to be honest, I just copied the process from there. And love uniformity in a project

File: 2.process-data/split_training_test.py (hunk 3 above)
Comment (wagenrace) [reply 2.process-data/split_training_test.py]: Remove whole file. This is replaced with the notebooks in the same folder

Comment (wagenrace) [reply 2.process-data/split_training_test.py]: Remove whole file. This is replaced with the notebooks in the same folder

**Comment by wagenrace** [comment]
@mth-gsk and @gwaygenomics. I now done with the first version of random forest classification. Could you review it? Specially the results and future steps in the README.md

File: 3.random_forest/random-forest-pipeline.sh [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+#
+# Process the data using methods centered around random forest classifiers
+set -e
+
```
Comment (gwaygenomics) [inline 3.random_forest/random-forest-pipeline.sh]:
```suggestion

# First, convert all notebooks to scripts
jupyter nbconvert --to=script \
  --FilesWriter.build_directory=scripts/nbconverted \
  *.ipynb
```

This line will convert the notebook to a python script for easier commenting

File: 3.random_forest/README.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# random forest
```
Comment (gwaygenomics) [inline 3.random_forest/README.md]:
```suggestion
# Random Forest
```

File: 3.random_forest/README.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+Random forest, collection of random generated desicion trees, is a form of classifier. The classification always goes via a hard cut in a feature. This work well if all features are independed.
```
Comment (gwaygenomics) [inline 3.random_forest/README.md]:
```suggestion
A random forest, a collection of random generated decision trees, is a form of a classifier.
The classification always goes via a hard cut in a feature.
This work well if all features are independent.
```

I'm firmly in the one-sentence-per-line markdown camp - https://github.com/prettier/prettier/issues/3302

It makes markdown easier to review on github and doesn't change formatting

File: 3.random_forest/README.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+## naive attempt
```
Comment (gwaygenomics) [inline 3.random_forest/README.md]:
```suggestion
## Naive Attempt
```

File: 3.random_forest/README.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+Using [Sklearn Random Forest Classifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html) an estimator is trained on the [processed training data](../2.process-data/README.md).
```
Comment (gwaygenomics) [inline 3.random_forest/README.md]:
```suggestion
Using [Sklearn Random Forest Classifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html) we trained an estimator on the [processed training data](../2.process-data/README.md).
```

File: 3.random_forest/README.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+### parameters
```
Comment (gwaygenomics) [inline 3.random_forest/README.md]:
```suggestion
### Parameters

```

File: 3.random_forest/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+- n_estimators=1000 **(differ from default)**
```
Comment (gwaygenomics) [inline 3.random_forest/README.md]: Love the detail!

File: 3.random_forest/README.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+- min_impurity_split=None
+- bootstrap=True
+- oob_score=False
+- n_jobs=None
+- random_state=None
+- verbose=0
+- warm_start=False
+- class_weight=None
+
+### results
```
Comment (gwaygenomics) [inline 3.random_forest/README.md]:
```suggestion
### Results

```

File: 3.random_forest/README.md [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+The balance f1-score on the validation set is [0.328](scripts/html/random_forest.html). This is very low, but what is interesting is the [confusion matrix](scripts/html/random_forest.html). The prediction are dominated by 3 classes: dopaminereceptor, EGFR, ROCK. When looking at the [most common classes](../2.process-data/results/target_counts.tsv) we find the same 3 classes.
```
Comment (gwaygenomics) [inline 3.random_forest/README.md]:
```suggestion
The balanced f1-score on the validation set is [0.328](scripts/html/random_forest.html).
This is very low, but what is interesting is the [confusion matrix](scripts/html/random_forest.html).
The predictions are dominated by 3 classes: dopaminereceptor, EGFR, ROCK.
When looking at the [most common classes](../2.process-data/results/target_counts.tsv) we find the same 3 classes.
```

File: 3.random_forest/README.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+When we look at the next 3 most common classes most common class (adrenoceptor, DNA_intercalation, AMPA) we see very few predictions. Almost all of these images are labeled as one of the 3 dominated classes.
+There are a few classes (e.q. Ca2, Cdc25, eNos, rac1) are well predicted and seems to be easy cases.
+
+### future steps
```
Comment (gwaygenomics) [inline 3.random_forest/README.md]:
```suggestion
### Future Steps

```

File: 3.random_forest/README.md [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+- [ ] Balance training data. The model seems to favor classes that are very dominate.  
+- [ ] parameter optimization. Try different parameters to find best solution
```
Comment (gwaygenomics) [inline 3.random_forest/README.md]:
```suggestion
- [ ] Parameter optimization. Try different parameters to find best solution
```

**Comment by gwaygenomics** [comment]
I think this PR is in need of a quick rebase! There is some nice [rebase documentation here](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request).

Here are some steps that might help:

```bash
# Figure out the commit hash at the upstream master branch
# (Make sure upstream is configured as a remote branch pointed to cytodata/single-cell-classifier)
git merge-base random_forest upstream/master

# Copy and paste the hash output
git rebase -i <$HASH>
```

This will send you into your text editor.
From here, you can edit commits

We should drop the following commits since they are duplicates of #12:

* 091a31e2b37771d457a719925a14c4f3745a3135 (this commit also includes training and test data before processing, which we currently have in the `.gitignore` 2c5589fd3caf91b72b42fac0b043be994ffdfc75)
* 4878294f6e7167bb9f454fabddd9730856c2ada3
* c28018e6395381c3c6ce21bd2bebdd550a4e7784
* d329fb629ba203f71dd0c55f916f0963d21ad749

After the rebase is complete (make sure not to delete commits you actually want!) run:

```bash
# force push (really important not to delete stuff you want b/c this rewrites history)
git push -f
```

File: 3.random_forest/README.md (hunk 6 above)
Comment (wagenrace) [reply 3.random_forest/README.md]: I can agree with that camp

File: 3.random_forest/random-forest-pipeline.sh (hunk 4 above)
Comment (wagenrace) [reply 3.random_forest/random-forest-pipeline.sh]: But that script is not used or run anywhere. Is it the idea no notebooks are present in the final product?

File: 3.random_forest/README.md (hunk 10 above)
Comment (wagenrace) [reply 3.random_forest/README.md]: Thanks ☺

File: 3.random_forest/random-forest-pipeline.sh (hunk 4 above)
Comment (gwaygenomics) [reply 3.random_forest/random-forest-pipeline.sh]: notebooks are one of the major final products!

All files in `scripts/nbconverted` are just so its easy to view the actual python code _without_ the figures and syntax of `.ipynb` files. Its tough to pick out the actual python (or R) code in the `.ipynb` files.

**Comment by gwaygenomics** [comment]
A couple of other comments before merging:

1. Lets call this module `3.train-ebimage` or something like that. I think `3.random-forest` is too specific - we should leave room for other models (especially since random forests will only serve as an important baseline)
2. Make sure to `restart kernel` and `run all` in the jupyter notebook files. The notebooks should, ideally, all start at execution cell 1 and be executed sequentially. Doing this ensures that no variables were mistakenly referred to out of order. This is a big problem for future users.

**Comment by wagenrace** [comment]
@gwaygenomics  why the rebase? This branch is completely up to data with the current master branch

File: 3.train_ebimage/README.md [hunk 15]
```diff
@@ -0,0 +1,6 @@
+# Train EBImage
+In here are the algorithm using only the features of EBImage.
```
Comment (gwaygenomics) [inline 3.train_ebimage/README.md]:
```suggestion

In here are the algorithm using only the features of EBImage.
```

New paragraph line after headings

File: 3.train_ebimage/1.random_forest/random-forest-pipeline.sh [hunk 16]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+#
+# Process the data using methods centered around random forest classifiers
+set -e
+
```
Comment (gwaygenomics) [inline 3.train_ebimage/1.random_forest/random-forest-pipeline.sh]: Highly recommend converting `.ipynb` to `.py` and including _both_ files

File: 3.train_ebimage/1.random_forest/random-forest-pipeline.sh [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+# Run the notebook
+jupyter nbconvert --to=html \
+  --FilesWriter.build_directory=scripts/html \
+  --ExecutePreprocessor.kernel_name=python \
+  --ExecutePreprocessor.timeout=10000000 \
+  --execute random_forest.ipynb
```
Comment (gwaygenomics) [inline 3.train_ebimage/1.random_forest/random-forest-pipeline.sh]:
```suggestion
  --execute random_forest.ipynb
  
```

Adding new line to remove "no new line at bottom of file" indicator

**Comment by gwaygenomics** [comment]
> @gwaygenomics why the rebase? This branch is completely up to data with the current master branch

A couple of files are still included in this PR that might not belong. 

1. Do you think we should have `2.process-data/data/test.tsv.gz` and `2.process-data/data/train.tsv.gz`? 
  * I can totally see an argument to include them (they include all EBimage features and not many more files will be bigger).
  * If you think they should be included, please remove their mention in the [`.gitignore`](https://github.com/cytodata/single-cell-classifier/blob/master/.gitignore#L47l_49).
2. The change in [`2.process-data/process-data-pipeline.sh`](https://github.com/cytodata/single-cell-classifier/pull/25/files#diff-d26ca781e67e88224f07738293a6d9cd) should be removed.

These two additional changes should be all that's needed before merging.

Rebasing is one way to do this. It is definitely harder to do, but makes things easier to view later on. A quick (and definitely acceptable) alternative is to:

1. Remove files and commit (git rm 2.process-data/data/test.tsv.gz) or commit an updated `.gitignore`
2. change `2.process-data/process-data-pipeline.sh` back to 100755 with:

```bash
chmod +x 2.process-data/process-data-pipeline.sh`
```

and then commit

**Comment by wagenrace** [comment]
@gwaygenomics I have solved the changes in 2.process_data the ugly way. I will look in to rebasing some day. Seems a bit complicated for just making the commits pretty

**Pull request merged**